File: packages/go/internal/seguranca/ataques/ataque_confusao_campos.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// AtaqueConfusaoCampos testa se rearranjos/deslocamentos dos campos do token
// (integridade[32] . ciphertext[n] . iv[16]) são aceitos.
type AtaqueConfusaoCampos struct{}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueConfusaoCampos) Nome() string {
	return "Confusão de campos do token (reordenação/deslocamento)"
}

type varianteCampos struct {
	nome  string
	bruto []byte
}

// concatenar junta vários buffers.
func concatenar(partes ...[]byte) []byte {
	total := 0
	for _, p := range partes {
		total += len(p)
	}
	out := make([]byte, 0, total)
	for _, p := range partes {
		out = append(out, p...)
	}
	return out
}

// Executar roda o ataque.
func (a *AtaqueConfusaoCampos) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	texto := "MENSAGEM_PARA_TESTE_DE_CAMPOS"
	token, err := alvo.Encrypt([]byte(texto))
	if err != nil {
		return seguranca.ResultadoAtaque{}, err
	}
	prefixo := alvo.Prefixo()
	decodificado, err := alvo.Base64urlDecode(token[len(prefixo):])
	if err != nil {
		return seguranca.ResultadoAtaque{}, err
	}
	campos := alvo.Decompor(decodificado)

	integridade := campos.Integridade
	ciphertext := campos.Ciphertext
	iv := campos.IV

	ivInvertido := append([]byte(nil), iv...)
	for i, j := 0, len(ivInvertido)-1; i < j; i, j = i+1, j-1 {
		ivInvertido[i], ivInvertido[j] = ivInvertido[j], ivInvertido[i]
	}

	variantes := []varianteCampos{
		{nome: "iv no início", bruto: concatenar(iv, integridade, ciphertext)},
		{nome: "ciphertext antes da integridade", bruto: concatenar(ciphertext, integridade, iv)},
		{nome: "iv duplicado no fim", bruto: concatenar(integridade, ciphertext, iv, iv)},
		{nome: "integridade encurtada", bruto: concatenar(integridade[1:], ciphertext, iv)},
		{nome: "byte extra no início", bruto: concatenar([]byte("X"), integridade, ciphertext, iv)},
		{nome: "byte extra entre integridade e ciphertext", bruto: concatenar(integridade, []byte("X"), ciphertext, iv)},
		{nome: "byte extra antes do iv", bruto: concatenar(integridade, ciphertext, []byte("X"), iv)},
		{nome: "iv rotacionado", bruto: concatenar(integridade, ciphertext, ivInvertido)},
		{
			nome:  "iv e último byte do ciphertext trocados",
			bruto: concatenar(integridade, ciphertext[:len(ciphertext)-1], iv, ciphertext[len(ciphertext)-1:]),
		},
	}

	aceitas := []string{}
	for _, v := range variantes {
		tokenVariante := prefixo + alvo.Base64urlEncode(v.bruto)
		if r, err := alvo.Decrypt(tokenVariante); err == nil {
			aceitas = append(aceitas, fmt.Sprintf("%s -> aceito (retornou %d bytes)", v.nome, len([]byte(r))))
		}
	}

	if len(aceitas) > 0 {
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevCritica,
			Detalhes:   fmt.Sprintf("%d rearranjo(s) de campo foram aceitos", len(aceitas)),
			Dados:      map[string]any{"exemplos": aceitas},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todos os %d rearranjos de campo foram rejeitados", len(variantes)),
	}, nil
}

File: packages/rust/src/seguranca/ataques/ataque_confusao_campos.rs
//! O token é integridade[32] . ciphertext[n] . iv[16]. Se o parser for
//! ambíguo, um atacante pode reordenar/deslocar os campos e construir um
//! token que sistemas diferentes interpretam de formas diferentes. Todos os
//! rearranjos devem ser rejeitados pela integridade.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::criptografia_alvo::CriptografiaAlvo;
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};

pub struct AtaqueConfusaoCampos;

impl Ataque for AtaqueConfusaoCampos {
    fn nome(&self) -> String {
        "Confusão de campos do token (reordenação/deslocamento)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        if alvo.as_any().downcast_ref::<CriptografiaAlvo>().is_none() {
            return Err(ErroAtaque::Skip(
                "Precisa de base64url_encode/decode do alvo.".to_string(),
            ));
        }

        let texto = "MENSAGEM_PARA_TESTE_DE_CAMPOS";
        let token = alvo
            .encrypt(texto.as_bytes())
            .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
        let prefixo = alvo.prefixo();
        let decodificado = alvo
            .base64url_decode(&token[prefixo.len()..])
            .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
        let campos = alvo.decompor(&decodificado);

        let integridade = &campos.integridade;
        let ciphertext = &campos.ciphertext;
        let iv = &campos.iv;

        let mut iv_rotacionado = iv.clone();
        iv_rotacionado.reverse();

        let variantes: Vec<(&str, Vec<u8>)> = vec![
            (
                "iv no início",
                [iv.clone(), integridade.clone(), ciphertext.clone()].concat(),
            ),
            (
                "ciphertext antes da integridade",
                [ciphertext.clone(), integridade.clone(), iv.clone()].concat(),
            ),
            (
                "iv duplicado no fim",
                [integridade.clone(), ciphertext.clone(), iv.clone(), iv.clone()].concat(),
            ),
            (
                "integridade encurtada",
                [integridade[1..].to_vec(), ciphertext.clone(), iv.clone()].concat(),
            ),
            (
                "byte extra no início",
                [b"X".to_vec(), integridade.clone(), ciphertext.clone(), iv.clone()].concat(),
            ),
            (
                "byte extra entre integridade e ciphertext",
                [integridade.clone(), b"X".to_vec(), ciphertext.clone(), iv.clone()].concat(),
            ),
            (
                "byte extra antes do iv",
                [integridade.clone(), ciphertext.clone(), b"X".to_vec(), iv.clone()].concat(),
            ),
            (
                "iv rotacionado",
                [integridade.clone(), ciphertext.clone(), iv_rotacionado].concat(),
            ),
            (
                "iv e último byte do ciphertext trocados",
                [
                    integridade.clone(),
                    ciphertext[..ciphertext.len() - 1].to_vec(),
                    iv.clone(),
                    ciphertext[ciphertext.len() - 1..].to_vec(),
                ]
                .concat(),
            ),
        ];

        let mut aceitas: Vec<String> = Vec::new();
        for (nome, bruto) in &variantes {
            let token_variante = format!("{}{}", prefixo, alvo.base64url_encode(bruto));
            if let Ok(r) = alvo.decrypt(&token_variante) {
                aceitas.push(format!(
                    "{} -> aceito (retornou {} bytes)",
                    nome,
                    r.len()
                ));
            }
        }

        if !aceitas.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Critica,
                &format!("{} rearranjo(s) de campo foram aceitos", aceitas.len()),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Todos os {} rearranjos de campo foram rejeitados",
                variantes.len()
            ),
        ))
    }
}
